rasterio/_err.pyx

```diff
@@ -39,43 +41,157 @@ cdef extern from "cpl_error.h":
+    pass
+
+
+class CPLE_OpenFailed(CPLError):
+    pass
+
+
+class CPLE_IllegalArg(CPLError):
+    pass
+
+
+class CPLE_NotSupported(CPLError):
+    pass
+
+
+class CPLE_AssertionFailed(CPLError):
+    pass
+
+
+class CPLE_NoWriteAccess(CPLError):
+    pass
+
+
+class CPLE_UserInterrupt(CPLError):
+    pass
+
+
+class ObjectNull(CPLError):
+    pass
+
+
+class CPLE_HttpResponse(CPLError):
+    pass
+
+
+class CPLE_AWSBucketNotFound(CPLError):
+    pass
+
+
+class CPLE_AWSObjectNotFound(CPLError):
+    pass
+
+
+class CPLE_AWSAccessDenied(CPLError):
+    pass
+
+
+class CPLE_AWSInvalidCredentials(CPLError):
+    pass
+
+
+class CPLE_AWSSignatureDoesNotMatch(CPLError):
+    pass
+
+
+# Map of GDAL error numbers to the Python exceptions.
 exception_map = {
-    1: RuntimeError,        # CPLE_AppDefined
-    2: MemoryError,         # CPLE_OutOfMemory
-    3: IOError,             # CPLE_FileIO
-    4: IOError,             # CPLE_OpenFailed
-    5: TypeError,           # CPLE_IllegalArg
-    6: ValueError,          # CPLE_NotSupported
-    7: AssertionError,      # CPLE_AssertionFailed
-    8: IOError,             # CPLE_NoWriteAccess
-    9: KeyboardInterrupt,   # CPLE_UserInterrupt
-    10: ValueError          # ObjectNull
-    }
-
-
-cdef class GDALErrCtxManager:
+    1: CPLE_AppDefined,
+    2: CPLE_OutOfMemory,
+    3: CPLE_FileIO,
+    4: CPLE_OpenFailed,
+    5: CPLE_IllegalArg,
+    6: CPLE_NotSupported,
+    7: CPLE_AssertionFailed,
+    8: CPLE_NoWriteAccess,
+    9: CPLE_UserInterrupt,
+    10: ObjectNull,
+
+    # error numbers 11-16 are introduced in GDAL 2.1. See 
+    # https://github.com/OSGeo/gdal/pull/98.
+    11: CPLE_HttpResponse,
+    12: CPLE_AWSBucketNotFound,
+    13: CPLE_AWSObjectNotFound,
+    14: CPLE_AWSAccessDenied,
+    15: CPLE_AWSInvalidCredentials,
+    16: CPLE_AWSSignatureDoesNotMatch}
+
+
+# CPL Error types as an enum.
+class GDALError(IntEnum):
+    none = 0    # CE_None
+    debug = 1   # CE_Debug
+    warning= 2  # CE_Warning
+    failure = 3 # CE_Failure
+    fatal = 4   # CE_Fatal
+
+
+cdef class CPLErrors:
     """A manager for GDAL error handling contexts."""
 
+    def check(self):
+        """Check the errror stack and raise or exit as appropriate."""
+        cdef const char *msg_c = NULL
+
+        err_type = CPLGetLastErrorType()
+        # Return True if there's no error.
+        # Debug and warnings are already picked up by the drivers()
+        # context manager.
+        if err_type < 3:
+            CPLErrorReset()
+            return
+
+        err_no = CPLGetLastErrorNo()
+        msg_c = CPLGetLastErrorMsg()
+        if msg_c == NULL:
+            msg = "No error message."
+        else:
+        # Reformat messages.
+            msg_b = msg_c
+            msg = msg_b.decode('utf-8')
+            msg = msg.replace("`", "'")
+            msg = msg.replace("\n", " ")
+
+        if err_type == 4:
+            sys.exit("Fatal error: {0}".format((err_type, err_no, msg)))
```

Comment:
A string is the same as 1, which I only recently learned. https://docs.python.org/2/library/sys.html#sys.exit